// @cloudconvert/n8n-nodes-cloudconvert/dist/nodes/CloudConvert/operations/command/execute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeCommand = executeCommand;
const n8n_workflow_1 = require("n8n-workflow");
const Utils_1 = require("../../Utils");
async function executeCommand() {
    const items = this.getInputData();
    const returnData = [];
    for (let i = 0; i < items.length; i++) {
        const engine = this.getNodeParameter('commandEngine', i);
        const commandBase = this.getNodeParameter('commandBase', i);
        const argumentsParam = this.getNodeParameter('commandArguments', i, '');
        const captureOutput = this.getNodeParameter('commandCaptureOutput', i, false);
        if (!engine) {
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Engine must be set for command operations');
        }
        if (!commandBase) {
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Command must be set for command operations');
        }
        const tasks = {
            'n8n-upload': {
                operation: 'import/upload',
            },
            'n8n-process': {
                input: 'n8n-upload',
                operation: 'command',
                engine,
                command: commandBase,
            },
            'n8n-export': {
                input: 'n8n-process',
                operation: 'export/url',
            },
        };
        if (argumentsParam) {
            tasks['n8n-process'].arguments = argumentsParam;
        }
        if (captureOutput) {
            tasks['n8n-process'].capture_output = true;
        }
        if (this.getNodeParameter('additionalOptions', i, null)) {
            let additionalOptions = this.getNodeParameter('additionalOptions', i);
            try {
                additionalOptions = JSON.parse(additionalOptions);
            }
            catch (exception) {
                throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Additional Options must be a valid JSON');
            }
            tasks['n8n-process'] = (0, Utils_1.mergeObjects)(tasks['n8n-process'], additionalOptions);
        }
        const createdJob = await Utils_1.createJob.call(this, tasks);
        const uploadTask = (0, Utils_1.getJobUploadTask)(createdJob);
        if (uploadTask) {
            await Utils_1.uploadInputFile.call(this, uploadTask, i);
        }
        const completedJob = await Utils_1.waitForJob.call(this, createdJob.id);
        const exportUrls = (0, Utils_1.getJobExportUrls)(completedJob);
        let baseJson = {};
        if (captureOutput) {
            const commandTask = completedJob.tasks.find((task) => task.operation === 'command' && task.status === 'finished');
            if (commandTask === null || commandTask === void 0 ? void 0 : commandTask.result) {
                baseJson = commandTask.result;
            }
        }
        if (exportUrls.length === 0) {
            returnData.push({
                json: baseJson,
                pairedItem: {
                    item: i,
                },
            });
        }
        else {
            for (const exportUrl of exportUrls) {
                returnData.push({
                    json: baseJson,
                    binary: {
                        data: await Utils_1.downloadOutputFile.call(this, exportUrl),
                    },
                    pairedItem: {
                        item: i,
                    },
                });
            }
        }
    }
    return this.prepareOutputData(returnData);
}
