// @cloudconvert/n8n-nodes-cloudconvert/dist/credentials/CloudConvertApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudConvertApi = void 0;
class CloudConvertApi {
    constructor() {
        this.name = 'cloudConvertApi';
        this.displayName = 'CloudConvert API';
        this.documentationUrl = 'https://cloudconvert.com/dashboard/api/v2/keys';
        this.icon = { light: 'file:../nodes/CloudConvert/cloudconvert.svg', dark: 'file:../nodes/CloudConvert/cloudconvert_dark.svg' };
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                description: 'The API Key can be generated on your dashboard. The n8n nodes requires the "user.read", "task.read" and "task.write" scopes.',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.cloudconvert.com/v2',
                url: '/users/me',
            },
        };
    }
}
exports.CloudConvertApi = CloudConvertApi;
