// @elevenlabs/n8n-nodes-elevenlabs/dist/nodes/elevenlabs/ElevenLabs.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElevenLabs = void 0;
const voice_1 = require("./Descriptions/voice");
const utils_1 = require("./Descriptions/utils");
const speech_1 = require("./Descriptions/speech");
class ElevenLabs {
    constructor() {
        this.description = {
            displayName: 'ElevenLabs',
            name: 'elevenLabs',
            icon: 'file:elevenlabs.svg',
            group: ['transform'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Interact with ElevenLabs API',
            defaults: {
                name: 'ElevenLabs',
            },
            usableAsTool: true,
            inputs: ["main"],
            outputs: ["main"],
            credentials: [
                {
                    name: 'elevenLabsApi',
                    required: true,
                },
            ],
            requestDefaults: {
                method: 'POST',
                baseURL: 'https://api.elevenlabs.io/v1',
                headers: {
                    'Content-Type': 'application/json',
                },
            },
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Voice',
                            value: 'voice',
                        },
                        {
                            name: 'Speech',
                            value: 'speech',
                        },
                    ],
                    default: 'voice',
                },
                ...voice_1.VoiceOperations,
                ...voice_1.VoiceFields,
                ...speech_1.SpeechOperations,
                ...speech_1.SpeechFields,
            ]
        };
        this.methods = {
            listSearch: utils_1.listSearch,
        };
    }
}
exports.ElevenLabs = ElevenLabs;
