// @cloudconvert/n8n-nodes-cloudconvert/dist/credentials/CloudConvertOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudConvertOAuth2Api = void 0;
const scopes = ['user.read', 'task.read', 'task.write'];
class CloudConvertOAuth2Api {
    constructor() {
        this.name = 'cloudConvertOAuth2Api';
        this.extends = ['oAuth2Api'];
        this.displayName = 'CloudConvert OAuth2 API';
        this.documentationUrl = 'https://cloudconvert.com/api/v2#authentication';
        this.icon = { light: 'file:../nodes/CloudConvert/cloudconvert.svg', dark: 'file:../nodes/CloudConvert/cloudconvert_dark.svg' };
        this.properties = [
            {
                displayName: 'Grant Type',
                name: 'grantType',
                type: 'hidden',
                default: 'authorizationCode',
            },
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: 'https://cloudconvert.com/oauth/authorize',
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: 'https://cloudconvert.com/oauth/token',
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: '',
            },
            {
                displayName: 'Authentication',
                name: 'authentication',
                type: 'hidden',
                default: 'body',
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: scopes.join(' '),
            },
        ];
    }
}
exports.CloudConvertOAuth2Api = CloudConvertOAuth2Api;
