// @cloudconvert/n8n-nodes-cloudconvert/dist/nodes/CloudConvert/operations/metadata/execute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeMetadata = executeMetadata;
const n8n_workflow_1 = require("n8n-workflow");
const Utils_1 = require("../../Utils");
async function executeMetadata() {
    var _a, _b;
    const items = this.getInputData();
    const returnData = [];
    for (let i = 0; i < items.length; i++) {
        const nodeOperation = this.getNodeParameter('operation', i);
        if (nodeOperation === 'metadata') {
            const tasks = {
                'n8n-upload': {
                    operation: 'import/upload',
                },
                'n8n-process': {
                    input: 'n8n-upload',
                    operation: 'metadata',
                },
            };
            const createdJob = await Utils_1.createJob.call(this, tasks);
            const uploadTask = (0, Utils_1.getJobUploadTask)(createdJob);
            if (uploadTask) {
                await Utils_1.uploadInputFile.call(this, uploadTask, i);
            }
            const completedJob = await Utils_1.waitForJob.call(this, createdJob.id);
            const metadata = (_b = (_a = completedJob.tasks.filter((task) => task.operation === 'metadata' && task.status === 'finished')[0]) === null || _a === void 0 ? void 0 : _a.result) === null || _b === void 0 ? void 0 : _b.metadata;
            returnData.push({
                json: metadata,
                pairedItem: {
                    item: i,
                },
            });
        }
        else if (nodeOperation === 'metadata/write') {
            const tasks = {
                'n8n-upload': {
                    operation: 'import/upload',
                },
                'n8n-process': {
                    input: 'n8n-upload',
                    operation: 'metadata/write',
                },
                'n8n-export': {
                    input: 'n8n-process',
                    operation: 'export/url',
                },
            };
            if (this.getNodeParameter('metadata', i, null)) {
                let metadataParam = this.getNodeParameter('metadata', i);
                try {
                    metadataParam = JSON.parse(metadataParam);
                }
                catch {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Metadata must be a valid JSON');
                }
                tasks['n8n-process'].metadata = metadataParam;
            }
            const createdJob = await Utils_1.createJob.call(this, tasks);
            const uploadTask = (0, Utils_1.getJobUploadTask)(createdJob);
            if (uploadTask) {
                await Utils_1.uploadInputFile.call(this, uploadTask, i);
            }
            const completedJob = await Utils_1.waitForJob.call(this, createdJob.id);
            const exportUrls = (0, Utils_1.getJobExportUrls)(completedJob);
            for (const exportUrl of exportUrls) {
                returnData.push({
                    json: {},
                    binary: {
                        data: await Utils_1.downloadOutputFile.call(this, exportUrl),
                    },
                    pairedItem: {
                        item: i,
                    },
                });
            }
        }
        else {
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Unsupported metadata operation: ${nodeOperation}`);
        }
    }
    return this.prepareOutputData(returnData);
}
