// @cloudconvert/n8n-nodes-cloudconvert/dist/nodes/CloudConvert/operations/pdf/execute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executePdfOperation = executePdfOperation;
const n8n_workflow_1 = require("n8n-workflow");
const Utils_1 = require("../../Utils");
async function executePdfOperation() {
    const items = this.getInputData();
    const returnData = [];
    for (let i = 0; i < items.length; i++) {
        const nodeOperation = this.getNodeParameter('operation', i);
        if (!nodeOperation.startsWith('pdf/')) {
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Unsupported PDF operation: ${nodeOperation}`);
        }
        const tasks = {
            'n8n-upload': {
                operation: 'import/upload',
            },
            'n8n-process': {
                input: 'n8n-upload',
                operation: nodeOperation,
            },
            'n8n-export': {
                input: 'n8n-process',
                operation: 'export/url',
            },
        };
        if (nodeOperation === 'pdf/a') {
            const conformanceLevel = this.getNodeParameter('pdfAConformanceLevel', i, '');
            if (conformanceLevel) {
                tasks['n8n-process'].conformance_level = conformanceLevel;
            }
        }
        else if (nodeOperation === 'pdf/ocr') {
            const autoOrient = this.getNodeParameter('pdfOcrAutoOrient', i, false);
            tasks['n8n-process'].auto_orient = autoOrient;
            const languages = this.getNodeParameter('pdfOcrLanguages', i, '');
            if (languages) {
                tasks['n8n-process'].language = languages
                    .split(',')
                    .map((l) => l.trim())
                    .filter((l) => l !== '');
            }
        }
        else if (nodeOperation === 'pdf/encrypt') {
            const setPassword = this.getNodeParameter('pdfEncryptSetPassword', i, '');
            if (setPassword) {
                tasks['n8n-process'].set_password = setPassword;
            }
            const setOwnerPassword = this.getNodeParameter('pdfEncryptSetOwnerPassword', i, '');
            if (setOwnerPassword) {
                tasks['n8n-process'].set_owner_password = setOwnerPassword;
            }
            const allowExtract = this.getNodeParameter('pdfEncryptAllowExtract', i, false);
            tasks['n8n-process'].allow_extract = allowExtract;
            const allowAccessibility = this.getNodeParameter('pdfEncryptAllowAccessibility', i, false);
            tasks['n8n-process'].allow_accessibility = allowAccessibility;
            const allowModify = this.getNodeParameter('pdfEncryptAllowModify', i, '');
            if (allowModify) {
                tasks['n8n-process'].allow_modify = allowModify;
            }
            const allowPrint = this.getNodeParameter('pdfEncryptAllowPrint', i, '');
            if (allowPrint) {
                tasks['n8n-process'].allow_print = allowPrint;
            }
        }
        else if (nodeOperation === 'pdf/decrypt') {
            const password = this.getNodeParameter('pdfDecryptPassword', i, '');
            if (password) {
                tasks['n8n-process'].password = password;
            }
        }
        else if (nodeOperation === 'pdf/split-pages') {
        }
        else if (nodeOperation === 'pdf/extract-pages') {
            const pages = this.getNodeParameter('pdfExtractPagesPages', i, '');
            if (pages) {
                tasks['n8n-process'].pages = pages;
            }
        }
        else if (nodeOperation === 'pdf/rotate-pages') {
            const pages = this.getNodeParameter('pdfRotatePagesPages', i, '');
            if (pages) {
                tasks['n8n-process'].pages = pages;
            }
            const rotation = this.getNodeParameter('pdfRotatePagesRotation', i, '');
            if (rotation) {
                tasks['n8n-process'].rotation = rotation;
            }
        }
        const createdJob = await Utils_1.createJob.call(this, tasks);
        const uploadTask = (0, Utils_1.getJobUploadTask)(createdJob);
        if (uploadTask) {
            await Utils_1.uploadInputFile.call(this, uploadTask, i);
        }
        const completedJob = await Utils_1.waitForJob.call(this, createdJob.id);
        const exportUrls = (0, Utils_1.getJobExportUrls)(completedJob);
        for (const exportUrl of exportUrls) {
            returnData.push({
                json: {},
                binary: {
                    data: await Utils_1.downloadOutputFile.call(this, exportUrl),
                },
                pairedItem: {
                    item: i,
                },
            });
        }
    }
    return this.prepareOutputData(returnData);
}
