// @cloudconvert/n8n-nodes-cloudconvert/dist/nodes/CloudConvert/operations/optimize/execute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeOptimize = executeOptimize;
const n8n_workflow_1 = require("n8n-workflow");
const Utils_1 = require("../../Utils");
const Utils_2 = require("../../Utils");
async function executeOptimize() {
    const items = this.getInputData();
    const returnData = [];
    for (let i = 0; i < items.length; i++) {
        const tasks = {
            'n8n-upload': {
                operation: 'import/upload',
            },
            'n8n-process': {
                input: 'n8n-upload',
                operation: 'optimize',
            },
            'n8n-export': {
                input: 'n8n-process',
                operation: 'export/url',
            },
        };
        if (this.getNodeParameter('additionalOptions', i, null)) {
            let additionalOptions = this.getNodeParameter('additionalOptions', i);
            try {
                additionalOptions = JSON.parse(additionalOptions);
            }
            catch (exception) {
                throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Additional Options must be a valid JSON');
            }
            tasks['n8n-process'] = (0, Utils_1.mergeObjects)(tasks['n8n-process'], additionalOptions);
        }
        const createdJob = await Utils_2.createJob.call(this, tasks);
        const uploadTask = (0, Utils_2.getJobUploadTask)(createdJob);
        if (uploadTask) {
            await Utils_2.uploadInputFile.call(this, uploadTask, i);
        }
        const completedJob = await Utils_2.waitForJob.call(this, createdJob.id);
        const exportUrls = (0, Utils_2.getJobExportUrls)(completedJob);
        for (const exportUrl of exportUrls) {
            returnData.push({
                json: {},
                binary: {
                    data: await Utils_2.downloadOutputFile.call(this, exportUrl),
                },
                pairedItem: {
                    item: i,
                },
            });
        }
    }
    return this.prepareOutputData(returnData);
}
