// @elevenlabs/n8n-nodes-elevenlabs/dist/credentials/ElevenLabsApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElevenLabsApi = void 0;
class ElevenLabsApi {
    constructor() {
        this.name = 'elevenLabsApi';
        this.displayName = 'ElevenLabs API';
        this.documentationUrl = 'https://docs.elevenlabs.io/api-reference';
        this.properties = [
            {
                displayName: 'ElevenLabs API Key',
                name: 'xiApiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'xi-api-key': '={{$credentials.xiApiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.elevenlabs.io/v2',
                url: '/voices',
            },
        };
    }
}
exports.ElevenLabsApi = ElevenLabsApi;
