// @cloudconvert/n8n-nodes-cloudconvert/dist/nodes/CloudConvert/operations/merge/execute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeMerge = executeMerge;
const Utils_1 = require("../../Utils");
async function executeMerge() {
    const items = this.getInputData();
    const returnData = [];
    const tasks = {};
    for (let i = 0; i < items.length; i++) {
        tasks['n8n-upload-' + i.toString()] = {
            operation: 'import/upload',
        };
    }
    tasks['n8n-process'] = {
        input: Object.keys(tasks),
        operation: 'merge',
        output_format: this.getNodeParameter('outputFormat', 0),
    };
    tasks['n8n-export'] = {
        input: 'n8n-process',
        operation: 'export/url',
    };
    const createdJob = await Utils_1.createJob.call(this, tasks);
    for (let i = 0; i < items.length; i++) {
        const uploadTask = (0, Utils_1.getJobUploadTask)(createdJob, i);
        if (uploadTask) {
            await Utils_1.uploadInputFile.call(this, uploadTask, i);
        }
    }
    const completedJob = await Utils_1.waitForJob.call(this, createdJob.id);
    const exportUrl = (0, Utils_1.getJobExportUrls)(completedJob)[0];
    returnData.push({
        json: {},
        binary: {
            data: await Utils_1.downloadOutputFile.call(this, exportUrl),
        },
    });
    return this.prepareOutputData(returnData);
}
