// @cloudconvert/n8n-nodes-cloudconvert/dist/nodes/CloudConvert/Utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createJob = createJob;
exports.uploadInputFile = uploadInputFile;
exports.getJobErrorMessage = getJobErrorMessage;
exports.waitForJob = waitForJob;
exports.downloadOutputFile = downloadOutputFile;
exports.getJobUploadTask = getJobUploadTask;
exports.getJobExportUrls = getJobExportUrls;
exports.mergeObjects = mergeObjects;
const n8n_workflow_1 = require("n8n-workflow");
async function createJob(tasks) {
    var _a, _b;
    const credentialsType = this.getNodeParameter('authentication', 0) === 'oAuth2'
        ? 'cloudConvertOAuth2Api'
        : 'cloudConvertApi';
    let createResponseData;
    try {
        createResponseData = await this.helpers.httpRequestWithAuthentication.call(this, credentialsType, {
            method: 'POST',
            url: 'https://api.cloudconvert.com/v2/jobs',
            json: true,
            body: {
                tag: 'n8n',
                tasks,
            },
        });
    }
    catch (e) {
        if ((_b = (_a = e.cause.response) === null || _a === void 0 ? void 0 : _a.data) === null || _b === void 0 ? void 0 : _b.code) {
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `${e.cause.response.data.message} (Code: ${e.cause.response.data.code})`);
        }
        throw e;
    }
    return createResponseData.data;
}
async function uploadInputFile(uploadTask, itemIndex = 0) {
    var _a, _b;
    const formData = new FormData();
    for (const parameter in ((_b = (_a = uploadTask.result) === null || _a === void 0 ? void 0 : _a.form) === null || _b === void 0 ? void 0 : _b.parameters) || {}) {
        formData.append(parameter, uploadTask.result.form.parameters[parameter]);
    }
    if (this.getNodeParameter('inputBinaryData', itemIndex)) {
        const binaryPropertyName = this.getNodeParameter('inputBinaryPropertyName', itemIndex);
        const binaryData = this.helpers.assertBinaryData(itemIndex, binaryPropertyName);
        const buffer = await this.helpers.getBinaryDataBuffer(itemIndex, binaryPropertyName);
        if (!binaryData.fileName)
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'No file name given for input file.');
        const blob = new Blob([buffer], { type: binaryData.mimeType || undefined });
        formData.append('file', blob, binaryData.fileName);
    }
    else {
        const content = this.getNodeParameter('inputFileContent', itemIndex);
        const filename = this.getNodeParameter('inputFilename', itemIndex);
        const blob = new Blob([content], { type: 'text/plain' });
        formData.append('file', blob, filename);
    }
    await this.helpers.httpRequest({
        method: 'POST',
        url: uploadTask.result.form.url,
        body: formData,
    });
}
function getJobErrorMessage(job) {
    return job.tasks
        .map((task) => {
        var _a;
        if (task.status === 'error' && task.code !== 'INPUT_TASK_FAILED') {
            return task.message + ' (Code: ' + ((_a = task.code) !== null && _a !== void 0 ? _a : '?') + ')';
        }
        return null;
    })
        .filter((msg) => msg !== null)
        .join('; ');
}
async function waitForJob(id) {
    const credentialsType = this.getNodeParameter('authentication', 0) === 'oAuth2'
        ? 'cloudConvertOAuth2Api'
        : 'cloudConvertApi';
    const waitResponseData = await this.helpers.httpRequestWithAuthentication.call(this, credentialsType, {
        method: 'GET',
        url: `https://sync.api.cloudconvert.com/v2/jobs/${id}`,
    });
    const job = waitResponseData.data;
    if (job.status === 'error') {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), getJobErrorMessage(job));
    }
    return job;
}
async function downloadOutputFile(exportUrl) {
    const downloadResponse = (await this.helpers.httpRequest({
        method: 'GET',
        url: exportUrl.url,
        returnFullResponse: true,
        encoding: 'stream',
    }));
    return this.helpers.prepareBinaryData(downloadResponse.body, exportUrl.filename, downloadResponse.headers['content-type']);
}
function getJobUploadTask(job, index = 0) {
    const uploadTasks = job.tasks.filter((task) => task.operation === 'import/upload');
    return uploadTasks[index] !== undefined ? uploadTasks[index] : null;
}
function getJobExportUrls(job) {
    return job.tasks
        .filter((task) => task.operation === 'export/url' && task.status === 'finished')
        .flatMap((task) => { var _a, _b; return (_b = (_a = task.result) === null || _a === void 0 ? void 0 : _a.files) !== null && _b !== void 0 ? _b : []; });
}
function mergeObjects(target, source) {
    const output = { ...target };
    for (const key of Object.keys(source)) {
        const sourceValue = source[key];
        const targetValue = output[key];
        if (sourceValue &&
            typeof sourceValue === 'object' &&
            !Array.isArray(sourceValue) &&
            targetValue &&
            typeof targetValue === 'object' &&
            !Array.isArray(targetValue)) {
            output[key] = mergeObjects(targetValue, sourceValue);
        }
        else {
            output[key] = sourceValue;
        }
    }
    return output;
}
